core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2080,10 +2080,24 @@ private DocumentMapperForType docMapper(String type) {
 
     private EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
         Sort indexSort = indexSortSupplier.get();
+        final boolean forceNewHistoryUUID;
+        switch (shardRouting.recoverySource().getType()) {
+            case EMPTY_STORE:
```

Comment:
`EMPTY_STORE` will always result in a new history uuid as we create a new index writer and it will not have any user data. I'm fine with setting `forceNewHistoryUUID` to true to be explicit (and later on morph this into just "generateHistoryUUID` once all existing indices are known to have one and we can be explicit about the moments we generate new ones).

Re forcing a stale allocations - good catch. As you say, that's a more complicated one indeed. I think we should do it as a follow up and discuss possible approaches. I'm thinking about a new recovery source. I think this is an edge case we should be explicit about.